lib/rules/only-extend-defined.js

```diff
@@ -0,0 +1,121 @@
+'use strict';
+
+// Traverse the style object and gather all possible paths
+const gatherDefinedStylePaths = (acc, styleObject, currentPath = '') => {
+  styleObject.properties.forEach((property) => {
+    if (property.computed) {
+      // Skip over computed properties for now.
+      // e.g. `{ [foo]: { ... } }`
+      // TODO handle this better?
+      return;
+    }
+
+    if (property.type === 'ExperimentalSpreadProperty' || property.type === 'SpreadElement') {
+      // Skip over object spread for now.
+      // e.g. `{ ...foo }`
+      // TODO handle this better?
+      return;
+    }
+
+    const propertyName = property.key.value || property.key.name;
+    const path = currentPath ? `${currentPath}.${propertyName}` : propertyName;
+
+    // Save the path
+    acc[path] = true;
+
+    // Continue traversing
+    if (property.value.type === 'ObjectExpression') {
+      gatherDefinedStylePaths(acc, property.value, path);
+    }
+  });
+};
+
+// Traverse extendable styles and validate that all possible paths exist in the defined styles
+const checkExtendedStyles = (context, definedStylePaths, extendedStyle, currentPath = '') => {
+  extendedStyle.properties.forEach((property) => {
+    if (property.computed) {
+      // Skip over computed properties for now.
+      // e.g. `{ [foo]: { ... } }`
```

Comment:
I think this raises an interesting question: should there be a rule to disallow computed properties for extendable styles (or for styles in general)?

Reply:
I've seen a lot of usages of constants for things around responsiveness (e.g. `[responsive.mediumAndAbove]`). I think we would begin to limit ourselves a lot if we disallowed computed properties.

I wonder if we can ensure that the same variable is used in both places... I can take a look tonight.

Reply:
Ah right! Can we add some valid and invalid test cases for nested styles? Media queries, `@supports` queries, pseudo-elements, pseudo-selectors, etc.